apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/CorrelationContext.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.skywalking.apm.agent.core.context;
+
+import org.apache.skywalking.apm.agent.core.base64.Base64;
+import org.apache.skywalking.apm.agent.core.conf.Config;
+import org.apache.skywalking.apm.util.StringUtil;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Correlation context, use to propagation user custom data.
+ * Working on the protocol and delegate set/get method.
+ */
+public class CorrelationContext {
+
+    private final Map<String, String> data;
+
+    public CorrelationContext() {
+        this.data = new HashMap<>(Config.Correlation.ELEMENT_MAX_NUMBER);
+    }
+
+    public Optional<String> set(String key, String value) {
+        // key must not null
+        if (key == null) {
+            return Optional.empty();
+        }
+        if (value == null) {
+            value = "";
+        }
```

Comment:
Do you mean delete the key when the value is null? Should I add a new method to implement the `remove` method?

Reply:
> Do you mean delete the key when the value is null? Should I add a new method to implement the `remove` method?

Adding a new API `remove` looks good to me, but I prefer to remove the item when the `value == null`(or even when `value.isEmpty()`) because the places in the context are "precious", and propagating `null` or `""` is a waste of bandwidth, WDYT @wu-sheng

Reply:
Agree, we should remove the element when value doesn't exist. And for `remove`, adding or not both works for me. But the logic of `null/empty` triggers `removing` should be documented clearly.

Reply:
@mrproliu please perform a removal when `value == null || value.isEmpty()`, and document it, `remove` is unnecessary then (I tend to keep the core API as simple as possible)

Reply:
fixed. It will remove the item when the value is empty or null.